Added BT HC05 module

coder137 opened the pull request:
- Added low level read and write functionality
- Updated main.cpp with btserial

coder137 committed:
Added BT HC05 module

- Added low level read and write functionality
- Updated main.cpp with btserial

ganeshram93 commented on HelloWorld_ST_Sensors/main.cpp:
```diff
@@ -39,34 +39,50 @@
+static spec::CarbonMonoxide co(PA_0, PA_1);                         // UART4
+static sensor::SGP30 sgp30(PB_9, PB_8);                             // I2C1
+static i2c_slave::SlaveCommunication slave(PC_1, PC_0, co, sgp30);  // I2C3
+static bt::hc05 btserial(PA_2, PA_3, 9600);                         // UART2
```
Can we define the baud rate value as a const/macro inside hc05.hpp?

coder137 replied:
This hasn't been done since we can configure our Bluetooth module to have different baud rates.
Since it is user defined

ganeshram93 commented on HelloWorld_ST_Sensors/user/bt/hc05.hpp:
```diff
@@ -0,0 +1,33 @@
+  static constexpr uint8_t kBufferSize = 200;
+
+ public:
+  hc05(PinName tx, PinName rx, uint32_t baud) : bt_uart_(tx, rx) {
```
Is there a reason why we are not passing "baud" as a bt_uart_ constructor's parameter or either way is the same?

coder137 replied:
Added to constructor

ganeshram93 commented on HelloWorld_ST_Sensors/user/bt/hc05.hpp:
```diff
@@ -0,0 +1,33 @@
+ public:
+  hc05(PinName tx, PinName rx, uint32_t baud) : bt_uart_(tx, rx) {
+    bt_uart_.baud(baud);
+    bt_uart_.attach(callback(this, &hc05::interrupt_cb));
```
Don't we need to pass "IrqType" inside the attach()?

ganeshram93 replied:
It looks like it defaults to RX interrupt, so the code looks good here.
A suggestion would be to explicitly mention it for quick clarity.

coder137 replied:
This is a bit of weirdness in the callback attach apis. Since I want to use a private class function the forums suggested I use this kind of API.

ganeshram93 commented on HelloWorld_ST_Sensors/user/bt/hc05.cpp:
```diff
@@ -0,0 +1,32 @@
+namespace bt {
+
+void hc05::write(const char* data) {
+  while (*data != 0) {
```
sprintf places null char at the end of the buffer.
But it's also better to safeguard with bounds checking, because the write() function is open-ended and it can be passed a char buffer not necessarily constructed using sprintf.

coder137 replied:
Added overloaded write method
i.e `write(const char * data, size_t bytes_to_write)`

ganeshram93 commented on HelloWorld_ST_Sensors/user/bt/hc05.cpp:
```diff
@@ -0,0 +1,32 @@
+
+namespace bt {
+
+void hc05::write(const char* data) {
```
Rename as bt_write_buff, for better clarity? (there is another write() API (asynchronous write) under RawSerial itself)

coder137 replied:
Since we are already using class based functionality, I did not make the functions verbose that way
For example

```cpp
bt::hc05 btserial(parameters);
btserial.write("My data");
```
As you can see the code is very readable for the user

coder137 replied:
NOTE: Also updated the API
```cpp
char buffer[] = "Hello World";
btserial.write(buffer, sizeof(buffer));
```

nikhilpeoria commented on HelloWorld_ST_Sensors/user/bt/hc05.cpp:
```diff
@@ -0,0 +1,32 @@
+
+void hc05::interrupt_cb() {
+  int data = bt_uart_.getc();
+  BUF_[counter_] = data;
```
Maybe use a more expressive name? The underscore is a little blink-and-you-miss-it

coder137 replied:
Changed `BUF_` name to `read_buf_`

coder137 committed:
Updated hc05

- Added write overloaded method for safer write over bluetooth
- Updated name for read buffer from BUF_ to read_buf_
- Added baud to RawSerial constructor
